### DISCUSSION TITLE
Compile Micropython with Ulab on SAMD51 Thing Plus

### DISCUSSION THREAD

**Discussion by LeandroFernandes88** [micropython/micropython#14017] [SAMD] [answered]
Hello everybody,

I'm looking for a way to compile Micropytohn with ulab in firmware for a teaching project. I tried to follow the guidelines in the Micropython and Ulab documents and forums as well. On the Sparkfun forum, I was recommended to use an AI to help with this task, but I was not successful either.

Would it be possible to compile the firmware this way (Micropython + Ulab) for SAMD51 Thing Plus? Do you know of any step-by-step instructions for this?
I'm new to this area, and I'm learning some things on my own.

I'm using Windows 11, and I tried the following steps:

I downloaded MSYS2 from the website: https://www.msys2.org/ file: msys2-x86_64-20240113.exe

I installed Python 3.11.8 and several other programs:
$ pacman -S python python-pip
$ pacman -S base-devel gcc
in addition to others suggested.

I cloned both repositories:
$ git clone https://github.com/micropython/micropython.git
$ git clone https://github.com/v923z/micropython-ulab.git

I copied the ulab code files (c and h) to the micropython extmod folder. To better organize it, I created a folder called "ulab" there.

I made some settings in the files, which I honestly don't understand:
mpconfigboard.h - in micropython\ports\samd
mpconfigboard.mk - in micropython\ports\samd\boards\SPARKFUN_SAMD51_THING_PLUS

and then perform the following commands:
export CROSS_COMPILE=arm-none-eabi-
$ make -C mpy-cross
$ make -C ports/samd BOARD=SPARKFUN_SAMD51_THING_PLUS

I sincerely apologize if the topic is not relevant, but I've been looking for Ulab in Micropython for a while and I haven't found any ready-made build (Micropython+Ulab) that I can use in SAMD51 Thing Plus and Teensy4.1

**Comment by rkompass** [comment, marked as answer]
I suggest trying to compile MP for your board without ULAB first. Without interfering (copying ULAB) in the cloned file structure.
Follow instructions [here](https://github.com/micropython/micropython/tree/master/ports/samd). 

> $ make -C mpy-cross
> This command should be executed from the root directory of this repository. All other commands below should be executed from the ports/stm32/ directory.

I assume that this last instruction was not corrected there. When doing the `make BOARD=SPARKFUN_SAMD51_THING_PLUS` you probably should be in the `ports/samd` MP subdirectory.

If you managed to do that and you find a nice firmware in the `boards/SPARKFUN_SAMD51_THING_PLUS` subdirectory  then try to make with
`make BOARD=SPARKFUN_SAMD51_THING_PLUS USER_C_MODULES=../../../ulab`  <-- specify the ULAB path relative to your current path.
Perhaps save the firmware binaries first and `make clean` and later compare. The firmware with ULAB should be larger.

**Reply by LeandroFernandes88** [reply]
I really appreciate the help,

I started using WSL2 on Windows and the process was quite simple.
I followed the instructions and everything worked fine (https://github.com/micropython/micropython/tree/master/ports/samd)

As for compiling the ulab, I still had some problems.
I cloned the repository in a folder in the side of the micropython folder
$ git clone https://github.com/v923z/micropython-ulab

Repeat the process of compiling the firmware
$ cd micropython
$ make -C mpy-cross
$ cd ports/samd
$ make BOARD=SPARKFUN_SAMD51_THING_PLUS submodules
$ make BOARD=SPARKFUN_SAMD51_THING_PLUS clean

and finally
$ make BOARD=SPARKFUN_SAMD51_THING_PLUS USER_C_MODULES=../../../micropython-ulab

I'm not sure if this is the intended way to compile micropython with ulab, but it seemed to work.

There was an error in the process and I asked for help for from the owner and maintainer of the ulab repository.
https://github.com/v923z/micropython-ulab/discussions/665

Thank you to everyone involved.

**Comment by v923z** [comment]
@LeandroFernandes88 Next time, please, do not cross-post! It's OK, if you mention an issue on multiple platforms, but use only links, if you want to do so. I've replied here https://github.com/v923z/micropython-ulab/discussions/665, but since you asked the same question here without pointing people to that post, no-one had an idea, what kind of progress you were making, or what you had already tried. 

By fragmenting efforts, you're doing a huge disservice to everyone involved, yourself included.

**Reply by LeandroFernandes88** [reply]
I am deeply sorry for this, and I sincerely apologize.

I'm not familiar with this type of firmware build operation.
I was using an operating system that wasn't suitable and I hadn't even managed to compile Micropython alone. Something that I imagine must be normal for experienced programmers.

I consider the ulab library very important and the efforts of everyone involved in making it accessible to Micropython and other IDEs fundamental for maintaining the ecosystem.

I just didn't know how to compile it as an additional module. And I had no idea if I was using the commands appropriately.

I probably imagine that most of the errors that have occurred are due to my inexperience in the matter.
